src/shared/mod.rs

```diff
@@ -0,0 +1,142 @@
+use crate::cache;
+use crate::protocol::notifications::{
+    create_diagnostics_notification, Notification,
+    PublishDiagnosticsParams,
+};
+use crate::protocol::properties::ContentChange;
+use crate::protocol::requests::{
+    Request, TextDocumentChangeParams, TextDocumentParams,
+    TextDocumentSaveParams,
+};
+use crate::utils;
+use crate::visitors::ast;
+
+pub fn parse_change_request(
+    data: String,
+) -> Result<Request<TextDocumentChangeParams>, String> {
+    let request: Request<TextDocumentChangeParams> =
+        Request::from_json(data.as_str())?;
+
+    Ok(request)
+}
+
+pub fn parse_save_request(
+    data: String,
+) -> Result<Request<TextDocumentSaveParams>, String> {
+    let request: Request<TextDocumentSaveParams> =
+        Request::from_json(data.as_str())?;
+
+    Ok(request)
+}
+
+pub fn parse_open_request(
+    data: String,
+) -> Result<Request<TextDocumentParams>, String> {
+    let request: Request<TextDocumentParams> =
+        Request::from_json(data.as_str())?;
+
+    Ok(request)
+}
+
+pub fn parse_close_request(
+    data: String,
+) -> Result<Request<TextDocumentParams>, String> {
+    let request: Request<TextDocumentParams> =
+        Request::from_json(data.as_str())?;
+
+    Ok(request)
+}
+
+pub fn apply_changes(
+    original: String,
+    changes: Vec<ContentChange>,
+) -> String {
+    for change in changes {
+        if change.range.is_none() {
```

Comment:
I realize you're just moving code, but I'm curious - when would range not be set for this message?

Reply:
Nvm, just found this in the docs: 'If range and rangeLength are omitted the new text is considered to be the full content of the document.`

Reply:
There is a server capacity called TextDocumentSyncKind in https://github.com/microsoft/language-server-protocol/blob/master/versions/protocol-2-x.md

if we enforce it to Full, we won't need to worry about the incremental changes. However, i still prefer to use the Incremental mode, since it's much less IO.

Reply:
Lets set it to full for now and add a card to move to incremental

Reply:
some of the lsp client ignore it anyway and always send full